### DISCUSSION TITLE
parsing list using ntc_templates

### DISCUSSION THREAD

**Discussion by henkq3** [networktocode/ntc-templates#1710] [General]
Hello, 

i'm trying to parse trkstat command from alcatel OXE equipment. I'm trying to use ntc_templates/textfsm libraries to parse it correctly. But i got trouble with list variables.

Here is an exemple of output:

+==============================================================================+
|           T R U N K    S T A T E     Trunk group number : 17                 |
|                                      Trunk group name   : 17-XXXX            |
|                                      Number of Trunks   : 30                 |
+------------------------------------------------------------------------------+
|   Index :     1    2    3    4    5    6    7    8    9   10   11   12   13  |
|   State :     F    F    F    B    F    B    F    F    B    F    B    F    B  |
+------------------------------------------------------------------------------+
|   Index :    14   15   16   17   18   19   20   21   22   23   24   25   26  |
|   State :     B    B    B    F    B    F    F    B    B    B    F    B    B  |
+------------------------------------------------------------------------------+
|   Index :    27   28   29   30                                               |
|   State :     F    F    F    B                                               |
+------------------------------------------------------------------------------+
|  F:  Free   |    B: Busy   |   Ct: busy Comp trunk   |  Cl: busy Comp link   |
|  WB: Busy Without B Channel|   Cr: busy Comp trunk for RLIO inter-ACT link   |
|  WBD: Data Transparency without chan.| WBM: Modem transparency without chan. |
|  D: Data Transparency      |   M: Modem transparency                         |
+------------------------------------------------------------------------------+

Remarks : it's possible to more/less index on each trunk: From 4 to 250 channels

I try to find inspiration from 'show vlan' with cisco_ios but cannot reproduce behaviour.

Here is my custom template:

`Value GROUP_NUMBER (\d+)
Value GROUP_NAME (\w+)
Value NUM_TRUNK (\d+)
Value List INDEX (\d+)
Value List STATE ([FBD])

Start
  ^\|\s+S I P\s+T R U N K\s+S T A T E\s+Trunk group number\s+:\s+${GROUP_NUMBER}
  ^\|\s+Trunk group name\s+:\s${GROUP_NAME}
  ^\|\s+Number of Trunks\s+:\s${NUM_TRUNK}
  ^\|\s+Index\s:\s+${INDEX}\s* -> Continue
  ^\s*{2}${INDEX}\s* -> Continue
  ^\s*{3}${INDEX}\s* -> Continue
  ^\s*{4}${INDEX}\s* -> Continue
  ^\s*{5}${INDEX}\s* -> Continue
  ^\s*{6}${INDEX}\s* -> Continue
  ^\s*{7}${INDEX}\s* -> Continue
  ^\s*{8}${INDEX}\s* -> Continue
  ^\s*{9}${INDEX}\s* -> Continue
  ^\s*{10}${INDEX}\s* -> Continue
  ^\s*{11}${INDEX}\s* -> Continue
  ^\s*{12}${INDEX}\s* -> Continue
  ^\s*{13}${INDEX}\s* -> Continue
  ^\|\s+State\s:\s+${STATE}\s* -> Continue
  ^\s*{2}${STATE}\s* -> Continue
  ^\s*{3}${STATE}\s* -> Continue
  ^\s*{4}${STATE}\s* -> Continue
  ^\s*{5}${STATE}\s* -> Continue
  ^\s*{6}${STATE}\s* -> Continue
  ^\s*{7}${STATE}\s* -> Continue
  ^\s*{8}${STATE}\s* -> Continue
  ^\s*{9}${STATE}\s* -> Continue
  ^\s*{10}${STATE}\s* -> Continue
  ^\s*{11}${STATE}\s* -> Continue
  ^\s*{12}${STATE}\s* -> Continue
  ^\s*{13}${STATE}\s* -> Continue`

I got error on line 13.

File "/usr/lib/python3.11/re/_parser.py", line 685, in _parse
    raise source.error("multiple repeat",
re.error: multiple repeat at position 7
...
textfsm.parser.TextFSMTemplateError: Invalid regular expression: '^\s+{2}(?P<INDEX>\d+)\s+*'. Line: 13.

Can someone help me to understand my error ?
Thanx

try debug, and change regex.
Try to apply advice from this page without succes : https://blog.networktocode.com/post/parsing-strategies-ntc-templates/

**Comment by henkq3** [comment]
Try to repost code sample in separated files :

[alcatel_oxe_trkstat.textfsm.txt](https://github.com/user-attachments/files/15544032/alcatel_oxe_trkstat.textfsm.txt)
[trkstat_cmd.txt](https://github.com/user-attachments/files/15544033/trkstat_cmd.txt)

**Comment by jorlandobr** [comment]
The "boxed" style seems to difficult a lot of capture. The extra spaces in TRUNKSTATE didn't helped also

Best I could figure out:

```
Value GROUP_NUMBER (\d+)
Value GROUP_NAME (\S+)
Value NUM_TRUNK (\d+)
Value List INDEX ((\d+\s)+)
Value List STATE ((\S+\s)+)

Start
  ^\+=.*				  
  ^\|\s+T\s+R\s+U\s+N\s+K\s+S\s+T\s+A\s+T\s+E\s+Trunk\s+group\s+number\s+:\s+${GROUP_NUMBER}
  ^\|\s+Trunk\s+group\s+name\s+:\s+${GROUP_NAME}
  ^\|\s+Number\s+of\s+Trunks\s+:\s+${NUM_TRUNK}
  ^\+-.*
  ^\|\s+Index\s+:\s+${INDEX}
  ^\|\s+State\s+:\s+${STATE}
  # Legend lines and end of output
  ^\|\s+F:.*
  ^\|\s+WB.*
  ^\|\s+D:.*
  # Error out if raw data does not match any above rules.
  ^.* -> Error "Could not parse line:"

```

**Comment by henkq3** [comment]
Hello @jorlandobr 

Thanx for the answer, the problem is not on the hearder, i can manage it
My problem is focused on list, when i try to apply your template, i only got first element of each line.

`[{'group_number': '17', 'group_name': '17-XXXX', 'num_trunk': '30', 'index': ['1 ', '14 ', '27 '], 'state': ['F ', 'B ', 'F ']}]
`

and i continue to try managing list in textfsm...

**Reply by jorlandobr** [reply]
@henkq3 ,
I couldn't reproduce your behavior.

My output is complete. I tested it at https://textfsm.nornir.tech/ and with a small python program.

My code:

```
import textfsm
# Load the input file to a variable
input_file = open("trkstat_alcatel_oxe.txt", encoding='utf-8')
raw_text_data = input_file.read()
input_file.close()

# Run the text through the FSM. 
# The argument 'template' is a file handle and 'raw_text_data' is a 
# string with the content from the show_inventory.txt file
template = open("trkstat_alcatel_oxe.textfsm")

re_table = textfsm.TextFSM(template)

fsm_results = re_table.ParseText(raw_text_data)
print (fsm_results)
```


My output:
`
PS C:\Users\python> python .\template_fsm.py
[['17', '17-XXXX', '30', ['1 2 3 4 5 6 7 8 9 10 11 12 13 ', '14 15 16 17 18 19 20 21 22 23 24 25 26 ', '27 28 29 30 '], ['F F F B F B F F B F B F B ', 'B B B F B F F B B B F B B ', 'F F F B ']]]
PS C:\Users\python>
`

Output at Nornir:
```
[
	{
		"GROUP_NAME": "17-XXXX",
		"GROUP_NUMBER": "17",
		"INDEX": [
			"1 2 3 4 5 6 7 8 9 10 11 12 13 ",
			"14 15 16 17 18 19 20 21 22 23 24 25 26 ",
			"27 28 29 30 "
		],
		"NUM_TRUNK": "30",
		"STATE": [
			"F F F B F B F F B F B F B ",
			"B B B F B F F B B B F B B ",
			"F F F B "
		]
	}
]
```

Maybe something with your code?  My textfsm version: 

```
pip list | findstr textfsm
textfsm                   1.1.3
```

**Comment by henkq3** [comment]
Hello @jorlandobr,

i haven't the your behaviour on my local machine and nornir textfsm page (please see screenshot) : only got the 1st reference of 
<img width="784" alt="test_textfsm_nornir" src="https://github.com/networktocode/ntc-templates/assets/85572439/114de425-2e8a-4eba-940e-8c0a1aa6e8b5">


On my local linux machine:`
pip list | grep -E "nornir|ntc|fsm|miko"
netmiko                   4.3.0
nornir                    3.4.1
ntc_templates             5.0.0
paramiko                  3.4.0
textfsm                   1.1.3

python -V
Python 3.11.2
`

With your template: 
`python template_fsm.py
[['17', '17-XXXX', '30', ['1 ', '14 ', '27 '], ['F ', 'B ', 'F ']]]`

And that not exactly what i need, i want to extract as list as it's possible interface list in [cisco_ios show vlan command](https://github.com/networktocode/ntc-templates/blob/master/ntc_templates/templates/cisco_ios_show_vlan.textfsm).
I try to reproduce the behaviour without success:

Instead of 
`"INDEX": [
    "1 2 3 4 5 6 7 8 9 10 11 12 13 ",
    "14 15 16 17 18 19 20 21 22 23 24 25 26 ",
    "27 28 29 30 "
]` need to have real list:`"INDEX": [
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30
]

"INDEX": [
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30"
]`
[trkstat_alcatel_oxe2.txt](https://github.com/user-attachments/files/15764371/trkstat_alcatel_oxe2.txt)

**Reply by mjbear** [reply]
> Instead of `"INDEX": [ "1 2 3 4 5 6 7 8 9 10 11 12 13 ", "14 15 16 17 18 19 20 21 22 23 24 25 26 ", "27 28 29 30 " ]` need to have real list:`"INDEX": [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30 ]
> 
> "INDEX": [ "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30" ]` [trkstat_alcatel_oxe2.txt](https://github.com/user-attachments/files/15764371/trkstat_alcatel_oxe2.txt)

@henkq3
So you're looking to carry over all the `INDEX` and `STATE` data one-by-one in a list, right?

Since this is from last June (2024) did you solve this?

**Comment by jorlandobr** [comment]
@henkq3,

My bad! I wasn't using your input text, I copy/pasted from your messages and lost the initial spaces.

At least it was easy to modify the template to give the needed output. See the lines with INDEX and STATE before and after the mofification.

```
Value GROUP_NUMBER (\d+)
Value GROUP_NAME (\S+)
Value NUM_TRUNK (\d+)
Value List INDEX ((\d+\s).*)
Value List STATE ((\S+\s).*)

Start
  ^\+=.*				  
  ^\|\s+T\s+R\s+U\s+N\s+K\s+S\s+T\s+A\s+T\s+E\s+Trunk\s+group\s+number\s+:\s+${GROUP_NUMBER}
  ^\|\s+Trunk\s+group\s+name\s+:\s+${GROUP_NAME}
  ^\|\s+Number\s+of\s+Trunks\s+:\s+${NUM_TRUNK}
  ^\+-.*
  ^\|\s+Index\s+:\s+${INDEX}\s+
  ^\|\s+State\s+:\s+${STATE}\s+
  # Legend lines and end of output
  ^\|\s+F:.*
  ^\|\s+WB.*
  ^\|\s+D:.*
  # Error out if raw data does not match any above rules.
  ^.* -> Error "Could not parse line:"
```
